### PR TITLE
Fix backbone freezing for question answering and speech recognition

### LINKED ISSUE #1112: Unable to generate predictions for wav2vec model fine-tuned with custom data
### Discussed in https://github.com/PyTorchLightning/pytorch-lightning/discussions/11432

<div type='discussions-op-text'>

<sup>Originally posted by **nayak24** January 11, 2022</sup>
Hi, I'm trying to fine-tune the baseline wav2vec model with my own audio training/test data using Lightning Flash, essentially exactly following the tutorial in this doc:
https://lightning-flash.readthedocs.io/en/latest/reference/speech_recognition.html 

However, I am running into an issue when generating the prediction for an audio file, and I'm getting a null output:

> 94.4 M    Trainable params
0         Non-trainable params
94.4 M    Total params
377.585   Total estimated model params size (MB)
Epoch 0: 100%|█████████████████████████████████████████████████████████| 88/88 [02:43<00:00,  1.86s/it, loss=633, v_num=57, train_loss_step=750.0]
Predicting: 88it [00:00, ?it/s]
[['']]

I'm not sure what the issue is, as I've only replaced the Timit dataset with my own input data for fine-tuning, and the rest of the script follows exactly from the doc above. All of the input data are wav files with the following format:

> format | 1 (uncompressed PCM)
number of channel | 1 (mono)
sampleRate | 16000
byteRate | 32000
blockAlign | 2
bitsPerSample (bit depth) | 16

I'm new to PyTorch Lightning and training with wav2vec as a whole, so I'm guessing that I'm missing something obvious. Any help would be greatly appreciated!

Here is the full script I'm running:

``` 
import torch
import flash
from flash.audio import SpeechRecognition, SpeechRecognitionData
from flash.core.data.utils import download_data

#download_data("https://pl-flash-data.s3.amazonaws.com/timit_data.zip", "./data")

datamodule = SpeechRecognitionData.from_csv(
    input_fields="file",
    target_fields="text",
    #train_file="data/timit/train.json",
    #test_file="data/timit/test.json",
    train_file="FLT034/FLT034-TRAIN.csv",
    test_file="FLT034/FLT034-TEST.csv",
    batch_size=4,
)

#can use any wav2vec model in HuggingFace as backbone for finetuning
model = SpeechRecognition(backbone="facebook/wav2vec2-base-960h")

#create trainer and finetune model
trainer = flash.Trainer(max_epochs=1)
trainer.finetune(model, datamodule=datamodule, strategy='no_freeze')

# predict on audio files
#datamodule = SpeechRecognitionData.from_files(predict_files=["data/timit/example.wav"], batch_size=4)
datamodule = SpeechRecognitionData.from_files(predict_files=["FLT034/FLT034-14.wav"], batch_size=4)
predictions = trainer.predict(model, datamodule=datamodule)
print(predictions)

# Save Checkpoint 
trainer.save_checkpoint("FL034_trained_model.pt") 
```

And here is a sample of the train.csv file with the annotations:
```
file,text
"./FLT034-12.wav","Weather at one seven five eight zulu."
"./FLT034-13.wav","Wind one niner zero at eight."
"./FLT034-14.wav","Visibility eight ceiling eight hundred overcast."
"./FLT034-15.wav","Temperature one five"
"./FLT034-16.wav","Dewpoint one four"
"./FLT034-17.wav","Altimeter three zero"
"./FLT034-18.wav","Get both sides on a mic"
```

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Fixes #1112

## Before submitting
- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the **[contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md)**, Pull Request section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes?
- [x] Did you write any **new necessary tests**? [not needed for typos/docs]
- [x] Did you verify **new and existing tests pass** locally with your changes?
- [ ] If you made a notable change (that affects users), did you **update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)**?

## PR review
 - [x] Is this pull request **ready for review**? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/text/question_answering/finetuning.py [hunk 1]
```diff
@@ -1,93 +0,0 @@
...
-        model = pl_module.transformer
-
-    elif model_type == "funnel":
-        model = pl_module.funnel
-
-    elif model_type == "xlnet":
-        model = pl_module.transformer
-    else:
-        # TODO: Is this the right way to exit the if statement ?
-        model = None
```
Comment (krshrimali) [inline flash/text/question_answering/finetuning.py:91]: So this is where the bug was?

File: flash/text/question_answering/model.py [hunk 2]
```diff
@@ -322,7 +321,7 @@ def _initialize_model_specific_parameters(self):
 
     def modules_to_freeze(self) -> Union[Module, Iterable[Union[Module, Iterable]]]:
         """Return the module attributes of the model to be frozen."""
-        return _get_question_answering_bacbones_for_freezing(self.model)
+        return self.model.base_model
```
Comment (krshrimali) [inline flash/text/question_answering/model.py:324]: Is there a chance that there is no attribute `base_model` and this will error out?

**Review by krshrimali — APPROVED** [review]
Nice fix! 🚀 Thanks, @ethanwharris! Just a couple of questions for my knowledge, but shouldn't block merging this PR.

File: flash/text/question_answering/finetuning.py [hunk 3]
```diff
@@ -1,93 +0,0 @@
...
-        elif model_type == "reformer":
-            model = pl_module.reformer
-
-    elif model_type in ["bart", "bigbird_pegasus", "led", "mbart"]:
-        if model_type in ["bart", "bigbird_pegasus", "mbart"]:
-            model = pl_module.model
-        elif model_type == "led":
-            model = pl_module.led
-
-    elif model_type == [
```
Comment (ethanwharris) [inline flash/text/question_answering/finetuning.py:33]: @krshrimali The equals sign here was a typo as `model_type` would be a string (should have been `in` for the question answering freezing to work correctly)

File: flash/text/question_answering/model.py (hunk 2 above)
Comment (krshrimali) [reply flash/text/question_answering/model.py:324]: Awesome, thanks Ethan!

File: flash/text/question_answering/finetuning.py (hunk 3 above)
Comment (krshrimali) [reply flash/text/question_answering/finetuning.py:33]: Got it! Good catch
